pkg/metrics/controller.go

```diff
@@ -0,0 +1,369 @@
+
+	ctx, _ := tag.New(context.Background(), tag.Upsert(keyName, fleetName))
+
+	c.recordWithTags(ctx, []tag.Mutator{tag.Upsert(keyType, "total")},
+		fleetsReplicasCountStats.M(int64(total)))
+	c.recordWithTags(ctx, []tag.Mutator{tag.Upsert(keyType, "allocated")},
+		fleetsReplicasCountStats.M(int64(allocated)))
+	c.recordWithTags(ctx, []tag.Mutator{tag.Upsert(keyType, "ready")},
+		fleetsReplicasCountStats.M(int64(ready)))
+	c.recordWithTags(ctx, []tag.Mutator{tag.Upsert(keyType, "desired")},
+		fleetsReplicasCountStats.M(int64(desired)))
+}
+
+func (c *Controller) recordWithTags(ctx context.Context, mutators []tag.Mutator, ms ...stats.Measurement) {
+	if err := stats.RecordWithTags(ctx, mutators, ms...); err != nil {
+		c.logger.WithError(err).Warn("error while recoding stats")
+	}
+}
+
+// recordGameServerStatusChanged records gameserver status changes, however since it's based
+// on cache events some events might collapsed and not appear, for example transition state
+// like creating, port allocation, could be skipped.
+// This is still very useful for final state, like READY, ERROR and since this is a counter
+// (as opposed to gauge) you can aggregate using a rate, let's say how many gameserver are failing
+// per second.
+// Addition to the cache are not handled, otherwise resync would make metrics inaccurate by doubling
+// current gameservers states.
+func (c *Controller) recordGameServerStatusChanges(old, new interface{}) {
+	newGs, ok := new.(*stablev1alpha1.GameServer)
+	if !ok {
+		return
+	}
+	oldGs, ok := old.(*stablev1alpha1.GameServer)
+	if !ok {
+		return
+	}
+	if newGs.Status.State != oldGs.Status.State {
+		fleetName := newGs.Labels[stablev1alpha1.FleetNameLabel]
+		if fleetName == "" {
+			fleetName = "none"
+		}
+		c.recordWithTags(context.Background(), []tag.Mutator{tag.Upsert(keyType, string(newGs.Status.State)),
+			tag.Upsert(keyFleetName, fleetName)}, gameServerTotalStats.M(1))
+	}
+}
+
+// record fleet allocations total by watching cache changes.
+func (c *Controller) recordFleetAllocationChanges(old, new interface{}) {
+	newFa, ok := new.(*stablev1alpha1.FleetAllocation)
+	if !ok {
+		return
+	}
+	oldFa, ok := old.(*stablev1alpha1.FleetAllocation)
+	if !ok {
+		return
+	}
+	// fleet allocations are added without gameserver allocated
+	// but then get modified on successful allocation with their gameserver
+	if oldFa.Status.GameServer == nil && newFa.Status.GameServer != nil {
+		c.recordWithTags(context.Background(), []tag.Mutator{tag.Upsert(keyFleetName, newFa.Spec.FleetName)},
+			fleetAllocationTotalStats.M(1))
+	}
+}
+
+// Run the Metrics controller. Will block until stop is closed.
+// Collect metrics via cache changes and parse the cache periodically to record resource counts.
+func (c *Controller) Run(workers int, stop <-chan struct{}) error {
+	err := crd.WaitForEstablishedCRD(c.crdGetter, "fleets.stable.agones.dev", c.logger)
```

Comment:
Since we already do this in the fleet controller (check for the established crd), should we remove it from here? We likely don't need to do this twice.

Reply:
true ! I should get rid of this